### PR TITLE
Add warning against unexpected --cfg with --check-cfg

### PR DESCRIPTION
**Pull request by Urgau**
This PR adds a warning when an unexpected `--cfg` is specified but not in the specified list of `--check-cfg`.

This is the follow-up PR I mentioned in https://github.com/rust-lang/rust/pull/99519.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_session/src/config.rs [hunk 1]
```diff
@@ -1159,6 +1159,31 @@ impl CrateCheckConfig {
         self.fill_well_known_names();
         self.fill_well_known_values();
     }
+
+    pub fn warn_actual(&self, opts: &Options, cfg: &CrateConfig) {
+        for &(name, value) in cfg {
+            if let Some(names_valid) = &self.names_valid {
+                if !names_valid.contains(&name) {
+                    early_warn(
```
Comment (Urgau) [inline compiler/rustc_session/src/config.rs:1167]: Is there a better way to emit a warning than using `early_warn` (which I believe cannot be enable/disable/promoted to error/...) ?

Comment (petrochenkov) [reply compiler/rustc_session/src/config.rs:1167]: Command line options are not tied to AST nodes, so I guess crate root is the only reasonable place to put attributes that can allow/deny/etc this warning.
Lint with a node can be added with `buffer_lint` as soon as `ParseSess` is available.

Comment (petrochenkov) [reply compiler/rustc_session/src/config.rs:1167]: I guess, the simplest solution would be to check this during the main early lint pass.

Comment (Urgau) [reply compiler/rustc_session/src/config.rs]: I used the previous lint and added to it a early lint pass implementation. I ended up not using `parse_sess` but directly using the methods on `EarlyContext`. I didn't even needed to use `CRATE_NODE_ID`.

**Comment by Urgau** [comment]
I've update the implementation to use a early lint pass. Let me know if there is something to change.

@rustbot ready

File: compiler/rustc_lint/src/builtin.rs [hunk 2]
```diff
@@ -3157,3 +3157,39 @@ impl<'tcx> LateLintPass<'tcx> for NamedAsmLabels {
         }
     }
 }
+
+use rustc_session::lint::builtin::UNEXPECTED_CFGS;
```
Comment (Urgau) [inline compiler/rustc_lint/src/builtin.rs:3161]: Should I create new lint for the (new ?) warning or re-using `UNEXPECTED_CFGS` is fine ?

Btw, this re-using is the reason why I needed to update `compiler/rustc_lint_defs/src/builtin.rs` and `src/librustdoc/core.rs`.

Comment (petrochenkov) [reply compiler/rustc_lint/src/builtin.rs:3161]: I think it's better to reuse `unexpected_cfgs` here.

File: src/test/ui/check-cfg/mix.stderr [hunk 3]
```diff
@@ -28,6 +28,14 @@ warning: unexpected `cfg` condition name
 LL | #[cfg_attr(uu, test)]
    |            ^^
 
+warning: unexpected `bar` as condition value for `feature` condition name
```
Comment (petrochenkov) [inline src/test/ui/check-cfg/mix.stderr:31]:
```suggestion
warning: unexpected condition value `bar` for condition name `feature`
```
The wording is a bit weird, maybe changing the word order will help.

**Comment by matthiaskrgr** [comment]
@bors r- looks like some tests failed in a rolluup
```
---- [ui] src/test/ui/check-cfg/mix.rs stdout ----
diff of stderr:

28	LL | #[cfg_attr(uu, test)]
29	   |            ^^
30	
-	warning: unexpected condition value `bar` for condition name `feature`
-	   |
-	   = help: was set with `--cfg` but isn't in the `--check-cfg` expected values
-	
35	warning: unexpected `unknown_name` as condition name
36	   |
37	   = help: was set with `--cfg` but isn't in the `--check-cfg` expected names

+	
+	warning: unexpected condition value `bar` for condition name `feature`
+	   |
+	   = help: was set with `--cfg` but isn't in the `--check-cfg` expected values
38	
39	warning: unexpected `cfg` condition name
40	  --> $DIR/mix.rs:35:10

````
https://github.com/rust-lang/rust/pull/101244

**Comment by Urgau** [comment]
I re-blessed the test output `src/test/ui/check-cfg/mix.rs`, should be good to go again.

@rustbot ready

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/101273#issuecomment-1234273790)

@bors r-

**Comment by Urgau** [comment]
:unamused: It's the same problem again. The order of the warnings changed again.

@petrochenkov Any idea how to make them not change order ?

**Comment by petrochenkov** [comment]
@Urgau 
Sort `cx.sess().parse_sess.config` before reporting, or use `FxIndexSet` instead of `FxHashSet` in it.

**Comment by Urgau** [comment]
I've rebased to fix the conflict in `compiler/rustc_lint_defs/src/builtin.rs` from https://github.com/rust-lang/rust/pull/94467.

@rustbot ready

**Comment by Urgau** [comment]
> == clock drift check ==
>   local time: Sat Sep  3 00:01:48 UTC 2022
> Session terminated, killing shell... ...killed.
> time="2022-09-03T00:01:48Z" level=error msg="error waiting for container: unexpected EOF"
>   network time: 
> Error: The operation was canceled.

Seems unrelated to the PR. Could be retried.

**Pull request merged**